src/SharedStream.mo

```diff
@@ -123,56 +125,58 @@ module {
     var concurrentChunksCounter : Nat = 0;
     public func sendChunk() : async* {
       #ok : Nat;
-      #err : { #Paused; #Busy; #SendChunkError : Text; #StreamClosed };
+      #err : ChunkError or { #Paused; #Busy; #SendChunkError : Text };
     } {
-      if (concurrentChunksCounter >= 5) {
+      if (concurrentChunksCounter >= maxConcurrentChunks_) {
         return #err(#Busy);
       };
       switch (lowestError) {
-        case (#Inf) {};
-        case (#Val _) { return #err(#Paused) };
+        case (null) {};
+        case (?le) { return #err(#Paused) };
       };
       let headId = queue.headIndex();
       streamIter.reset();
       let elements = Iter.toArray(streamIter);
-      if (elements.size() == 0) {
-        return #ok(0);
-      };
       try {
-        concurrentChunksCounter += 1;
-        let resp = await callback(streamId, elements, headId);
-        concurrentChunksCounter -= 1;
-        switch (resp) {
-          case (#err err) {
-            Timer.cancelTimer(heartbeatTimer);
-            return #err(
-              switch (err) {
-                case (#StreamClosed) #StreamClosed;
-                case (#NotRegistered) #SendChunkError("Not registered");
-              }
-            );
+        // if last call was more than 20 seconds ago, send anyway (keep-alive)
+        if (elements.size() > 0 or (Time.now() - lastChunkTimestamp) > 20_000_000_000) {
+          concurrentChunksCounter += 1;
+          lastChunkTimestamp := Time.now();
+          let resp = await sendFunc(streamId, elements, headId);
+          concurrentChunksCounter -= 1;
+          switch (resp) {
+            case (#err err) switch (err) {
+              case (#StreamClosed len) {
+                queue.pruneTo(len);
+                restoreHistoryIfNeeded();
+                return #err(#StreamClosed(len));
+              };
+              case (#NotRegistered) return #err(#SendChunkError("Not registered"));
+              case (#BrokenPipe _) return #err(#SendChunkError("Wrong index"));
+            };
+            case (#ok) {
+              queue.pruneTo(headId + elements.size());
+              restoreHistoryIfNeeded();
+            };
           };
-          case (#ok) lastChunkTimestamp := Time.now();
         };
-        queue.pruneTo(headId + elements.size());
-        restoreHistoryIfNeeded();
         #ok(elements.size());
       } catch (err : Error) {
         concurrentChunksCounter -= 1;
-        lowestError := #Val(switch (lowestError) { case (#Inf) headId; case (#Val val) Nat.min(val, headId) });
+        lowestError := ?(switch (lowestError) { case (null) headId; case (?val) Nat.min(val, headId) });
```

Comment:
`lowestError = do ? { Nat.min(headId, lowestError!) };`

Reply:
This line needs to set `headId` value in case `lowestError` is null